felix/calc/encapsulation_resolver.go

```diff
@@ -0,0 +1,224 @@
+		log.Debug("EncapsulationResolver: skip calculation because inSync is false")
+		return
+	}
+
+	if r.config != nil {
+		newIPIPEnabled := r.encapCalc.IPIPEnabled()
+		newVXLANEnabled := r.encapCalc.VXLANEnabled()
+
+		if r.config.Encapsulation.IPIPEnabled != newIPIPEnabled || r.config.Encapsulation.VXLANEnabled != newVXLANEnabled {
+			log.WithFields(log.Fields{
+				"oldIPIPEnabled":  r.config.Encapsulation.IPIPEnabled,
+				"newIPIPEnabled":  newIPIPEnabled,
+				"oldVXLANEnabled": r.config.Encapsulation.VXLANEnabled,
+				"newVXLANEnabled": newVXLANEnabled,
+			}).Info("EncapsulationResolver: Encapsulation changed. Restart Felix.")
+			r.configChangedRestartCallback()
+		}
+	}
+}
+
+// EncapsulationCalculator is a helper struct to aid in calculating if IPIP and/or VXLAN
+// encapsulation should be enabled based on the existing IP Pools and their
+// configuration. It is used by EncapsulationResolver in this file, where it watches for
+// encapsulation changes to restart Felix, and by Run() in daemon.go, where it calculates
+// the encapsulation state that will be effectively used by Felix.
+type EncapsulationCalculator struct {
+	config     *config.Config
+	ipipPools  map[string]struct{}
+	vxlanPools map[string]struct{}
+}
+
+func NewEncapsulationCalculator(config *config.Config, ippoolKVList *model.KVPairList) *EncapsulationCalculator {
+	encapCalc := &EncapsulationCalculator{
+		config:     config,
+		ipipPools:  map[string]struct{}{},
+		vxlanPools: map[string]struct{}{},
+	}
+
+	if ippoolKVList != nil {
+		encapCalc.initPools(ippoolKVList)
+	}
+
+	return encapCalc
+}
+
+func (c *EncapsulationCalculator) initPools(ippoolKVList *model.KVPairList) {
+	for _, kvp := range ippoolKVList.KVPairs {
+		err := c.handlePool(*kvp)
+		if err != nil {
+			log.Infof("error handling update %+v: %v. Ignoring.", *kvp, err)
+		}
+	}
+}
+
+func (c *EncapsulationCalculator) handlePool(p model.KVPair) error {
+	if _, ok := p.Key.(model.IPPoolKey); ok {
+		// When dealing with an model.IPPool, p.Value is nil for a removal
+		return c.handleModelPool(p)
+	}
+
+	if _, ok := p.Value.(*apiv3.IPPool); ok {
+		// When dealing with an apiv3.IPPool, p.Key is nil
+		return c.handleAPIPool(p)
+	}
+
+	return fmt.Errorf("Not a valid IP pool type")
+}
+
+func (c *EncapsulationCalculator) handleModelPool(p model.KVPair) error {
+	k, ok := p.Key.(model.IPPoolKey)
+	if !ok {
+		return fmt.Errorf("failed to convert %+v to model.IPPoolKey", p.Key)
+	}
+
+	poolKey := k.CIDR.String()
+	if p.Value == nil {
+		c.removePool(poolKey)
+	} else {
+		pool, ok := p.Value.(*model.IPPool)
+		if !ok {
+			return fmt.Errorf("failed to convert %+v to *model.IPPool", p.Value)
+		}
+		c.updatePool(poolKey, pool.IPIPMode != encap.Undefined, pool.VXLANMode != encap.Undefined)
+
+	}
+
+	return nil
+}
+
+func (c *EncapsulationCalculator) handleAPIPool(p model.KVPair) error {
+	if p.Value == nil {
+		return fmt.Errorf("API pool KVPair Value is nil")
```

Comment:
As this only comes from client.List() for IP pools, I couldn't figure out a deletion scenario (deletions are handled when we get them via the calc_graph, as `model.IPPool`)

Reply:
Probably still worth handling them, in case this ever receives a KVP from the syncer (if we say, switch to sending v3 updates over the syncer which is 100% on my wishlist)

Reply:
Good point, will add it! Thanks!

Reply:
Revisiting this: the issue is that (at least on the pools that are retrieved via API List()) the KVPair key is `nil`, so there is no way of getting the pool CIDR if the Value is also `nil`. I'll add handling of the case when the key is non-nil, but if both Key and Value are `nil` there's not much we can do...